### PR TITLE
MIGRATION-32 - Adding support for Recap Xblock.

### PR DESCRIPTION
**Pull request by Squirrel18**
## Description:

This PR adds support for Recap Xblock to the [CAPA](https://openedx.atlassian.net/wiki/spaces/AC/pages/128090267/Capa+Problem+Architecture) problem base.

## Previuos work:

https://github.com/proversity-org/edx-platform/commit/101db7795f89b4d8a04f5df8c21c5a3eff25aba0
https://github.com/proversity-org/edx-platform/commit/1e81c843e2a406c9935241467c0d00ba102b9c79 Not applied.
https://github.com/proversity-org/edx-platform/commit/a3f7acb5dff388733b1adb740181d017c32e3620
https://github.com/proversity-org/edx-platform/commit/82cce7774d1dd67dd243d2a1e788f812d1a8711b

## Reviewers:

- [ ] @andrey-canon 
- [ ] @diegomillan

### CODE REVIEW

File: common/lib/capa/capa/capa_problem.py [hunk 1]
```diff
@@ -217,6 +219,41 @@ def __init__(self, problem_text, id, capa_system, capa_module,  # pylint: disabl
...
                 self.extracted_tree = self._extract_html(self.tree)
 
+    def get_question_answer_text(self):
+        """
+        Get answer text from ChoiceResponse and MultipleChoiceResponse
+        types.
+        """
+        answer = ''
+        try:
+            if isinstance(self.responders.values()[0], responsetypes.ChoiceResponse):
```
Comment (andrey-canon) [inline common/lib/capa/capa/capa_problem.py]: `self.responders.values()` multiple use, you can assign it to a variable

File: common/lib/capa/capa/capa_problem.py [hunk 2]
```diff
@@ -217,6 +219,41 @@ def __init__(self, problem_text, id, capa_system, capa_module,  # pylint: disabl
...
+                        for k in val:
+                            if k == elem.get('name'):
+                                answer += "<div>{}</div>".format(elem.text)
+            elif isinstance(self.responders.values()[0], responsetypes.MultipleChoiceResponse):
+                for elem in self.responders.values()[0].get_choices():
+                    for val in self.student_answers.values():
+                        if val == elem.get('name'):
+                            answer += elem.text
+            return answer
+        except Exception as error:  # pylint: disable=broad-except
```
Comment (andrey-canon) [inline common/lib/capa/capa/capa_problem.py:243]: please be more specific

File: common/lib/capa/capa/responsetypes.py [hunk 3]
```diff
@@ -1027,6 +1030,22 @@ def setup_response(self):
...
+        """
+        Initialize name attributes in <choice> tags for this response.
+        """
+
+        for index, choice in enumerate(self.get_choices()):
+            choice.set("name", "choice_" + str(index))
+
+            # If a choice does not have an id, assign 'A' 'B', .. used by CompoundHint
+            if not choice.get('id'):
+                choice.set("id", chr(ord("A") + index))
```
Comment (andrey-canon) [inline common/lib/capa/capa/responsetypes.py]: double quotes

File: common/lib/xmodule/xmodule/capa_base.py [hunk 4]
```diff
@@ -1281,6 +1311,15 @@ def submit_problem(self, data, override_time=False):
         if not self.correctness_available():
             success = 'submitted'
 
+        try:
+            submission = self.lcp.get_question_answer_text()
+            self.student_answers_text = submission
+
+            if sub_api and self.runtime.get_real_user:
+                sub_api.create_submission(self.student_item_key(), submission)
+        except Exception as error:  # pylint: disable=broad-except
```
Comment (andrey-canon) [inline common/lib/xmodule/xmodule/capa_base.py:1320]: please be more specific

File: common/lib/capa/capa/capa_problem.py (hunk 2 above)
Comment (andrey-canon) [reply common/lib/capa/capa/capa_problem.py:243]: if the exception is raised, this method  will return nothing, so please be specific and return None or return the default value.

File: common/lib/capa/capa/capa_problem.py [hunk 5]
```diff
@@ -217,6 +219,41 @@ def __init__(self, problem_text, id, capa_system, capa_module,  # pylint: disabl
...
+        """
+        Get question text embedded in a <p> tag from xml element tree.
+        """
+        root = self.tree
+        question = ''
+        try:
+            for paragraph in root.iter('p'):
+                question += "<p>{}</p>".format(paragraph.text)
+            return question
+        except Exception as e:
```
Comment (andrey-canon) [inline common/lib/capa/capa/capa_problem.py]: same comment as above

**Review by andrey-canon — CHANGES_REQUESTED** [review]

File: common/lib/capa/capa/capa_problem.py (hunk 2 above)
Comment (Squirrel18) [reply common/lib/capa/capa/capa_problem.py:243]: Discussed.

File: common/lib/xmodule/xmodule/capa_base.py (hunk 4 above)
Comment (Squirrel18) [reply common/lib/xmodule/xmodule/capa_base.py:1320]: Discussed.

File: common/lib/capa/capa/capa_problem.py (hunk 5 above)
Comment (Squirrel18) [reply common/lib/capa/capa/capa_problem.py]: Discussed.

File: common/lib/capa/capa/capa_problem.py (hunk 1 above)
Comment (Squirrel18) [reply common/lib/capa/capa/capa_problem.py]: Addressed.

File: common/lib/capa/capa/responsetypes.py (hunk 3 above)
Comment (Squirrel18) [reply common/lib/capa/capa/responsetypes.py]: Addressed.

**Comment by Squirrel18** [comment]
PR addressed. @andrey-canon

**Review by andrey-canon — APPROVED** [review]

**Pull request merged**